Improve density and direction calculation in the neutral beam model

vsnever opened the pull request:
This updates density and direction calculation of the neutral beam model and fixes #414.

1. The beam deviation is now calculated as $$\sigma_x(z) = \sqrt{\sigma^2 + (ztg(\alpha_x))^2}\hspace{0.5cm}\sigma_y(z) = \sqrt{\sigma^2 + (ztg(\alpha_y))^2},$$ which is consistent with the Gaussian beam model, where the beam deviation σ(z) due to both beam divergence and deviation at beam origin is the result of the convolution of the two Gaussian functions as explained [here](https://github.com/cherab/core/issues/414#issuecomment-1549412429). Previously, the beam deviation was modelled as: $$\sigma_x(z) =\sigma + ztg(\alpha_x)\hspace{0.5cm}\sigma_y(z) = \sigma + ztg(\alpha_y),$$ which is unclear to me.
2. In this case, the (non-normalised) direction of velocity of the beam atoms can be calculated as follows: $$e_x = x\frac{(ztg(\alpha_x))^2}{\sigma^2 + (ztg(\alpha_x))^2},$$ $$e_y = y\frac{(ztg(\alpha_y))^2}{\sigma^2 + (ztg(\alpha_y))^2},$$ $$e_z = z.$$ This gives the following pattern for the beam direction (σ=0.1 m and α=5°):
![divergence](https://github.com/cherab/core/assets/22627041/79de06df-c16a-44e0-9754-f7cdb5320f51)

The broadening of the BES line shape is now naturally determined by the beam divergence. Therefore, setting beam temperature to the physically unreasonable values is no longer required. The parameters of the beam in the BES demo are updated.

Although this change does impact user results, I think it is justified by the improved consistency of the beam model.

vsnever committed:
Improve beam direction and beam density calculations.

vsnever committed:
Update BES demo and beam-related pictures in the docs.

Mateasek reviewed (changes requested):
Hi @vsnever,

thanks for the changes. The code itself looks clean and I couldn't find any problems. I would only suggest one improvement regarding the documentation. Since the beam direction became more complex, I think it would be nice to add the equations you use into the documentation. I think it would be enough if you add the equations and plot you posted in the opening post of this PR. It would be explanatory enough and would remove the necessity to go through the cython code for the users, what do you think?

jacklovell reviewed (changes requested):
Thanks @vsnever. I'm with @Mateasek: the changes are sound but it would be even better to explicitly document what calculation is being used.

I'd also suggest adding some emphasis (`**` or `*`) to the line in the changelog highlighting that this will change existing results.

If there aren't already any unit test associated with this, now is also a good time to add them. But don't worry if that's too big an project: I wouldn't want to hold up an otherwise useful change because of the work involved in setting up a mock beam for tests.

vsnever committed:
Added equations for Beam density and direction calculation. Moved SingleRayAttenuator from CXS models section to Beam Attenuation section.

vsnever committed:
Highlighted the line about beam density and direction change in CHANGELOG.md.

vsnever committed:
Added tests for the beam density and direction calculation.

vsnever committed:
Added beam direction plot to the docs.

vsnever commented:
Thanks for the reviews, @Mateasek, @jacklovell.

> Since the beam direction became more complex, I think it would be nice to add the equations you use into the documentation. I think it would be enough if you add the equations and plot you posted in the opening post of this PR. It would be explanatory enough and would remove the necessity to go through the cython code for the users, what do you think?

I added equations for the beam density and direction in `SIngleRayAttenuator.density()` and `Beam.direction()` docstrings respectively. I noticed that in the docs, the description of the `SingleRayAttenuator` class was placed under the "CXS models" section, so I moved it to a separate "Beam attenuation" section. The beam direction plot is added to the `particle_beams.rst`.

> I'd also suggest adding some emphasis (** or *) to the line in the changelog highlighting that this will change existing results.

I agree, the line announcing this change in the changelog is now in bold.

> If there aren't already any unit test associated with this, now is also a good time to add them. But don't worry if that's too big a project: I wouldn't want to hold up an otherwise useful change because of the work involved in setting up a mock beam for tests.

We already have tests for the beam CX model, so setting up a mock beam and plasma is not a problem. I added the tests for the beam density and direction. They test these values on-axis, off-axis and outside the beam.

vsnever committed:
Merge branch 'development' into enhancement/beam_direction_and_density

vsnever committed:
Placed the equations for the beam density and direction in the math block instead of inline.

jacklovell commented on cherab/core/beam/node.pyx:
```diff
@@ -232,7 +235,24 @@ cdef class Beam(Node):
 
     cpdef Vector3D direction(self, double x, double y, double z):
         """
```
Use `r"""` and get rid of double backslashes in math formulae.

vsnever replied:
Done.

jacklovell commented on cherab/core/model/attenuator/singleray.pyx:
```diff
@@ -87,17 +102,40 @@ cdef class SingleRayAttenuator(BeamAttenuator):
     @cython.cdivision(True)
     cpdef double density(self, double x, double y, double z) except? -1e999:
         """
```
`r"""` again.

vsnever replied:
Done,

jacklovell commented on cherab/core/tests/test_beam.py:
```diff
@@ -0,0 +1,156 @@
+        return self.value
+
+
+class TestAtomicData(AtomicData):
```
Rename this `MockAtomicData`: classes starting with `Test` are assumed by both unittest and pytest to contain test cases and this convention is broken here which in turn throws up warnings in the test discovery process.

jacklovell reviewed (changes requested):
Looks good. Couple of minor suggestions.

vsnever committed:
Converted docstrings to raw string format.

vsnever committed:
Rename TestAtomicData to MockAtomicData to avoid unittest warnings.

vsnever commented:
>Looks good. Couple of minor suggestions.

Thanks, @jacklovell. All done.

Mateasek reviewed (approved):
Thanks @vsnever for improving the docstrings, I have no further suggestions.
